### PR TITLE
More explicit error message if transport is already closed

### PR DESCRIPTION
**Pull request by lmazuel**
Fix https://github.com/Azure/azure-sdk-for-python/issues/25640

We're not sure if we can auto-reopen the connection once it's closed, this can lead to leaky socket behavior. Therefore, this PR is just making explicit with a better error message that it's closed already.

Interesting thing is that if we call `close` while we never opened it, it shouldn't fail (it would be a bad breaking change), so the approach here is to track the open and the close event, for better acuracy.

I take the opportunity of this PR to fix the typing cast, that was ironically hidding bug #25640 (mypy tried to tell me, but I didn't listen at the time :p)

### CODE REVIEW

**Comment by johanste** [comment]
We should make sure we didn't break child clients/shared transports (e.g. storage container -> blob client)

File: sdk/core/azure-core/azure/core/pipeline/transport/_aiohttp.py [hunk 1]
```diff
@@ -126,27 +129,35 @@ async def __aexit__(
...
-                "trust_env": self._use_env_settings,
-                "cookie_jar": jar,
-                "auto_decompress": False,
-            }
-            if self._loop is not None:
-                clientsession_kwargs["loop"] = self._loop
-            self.session = aiohttp.ClientSession(**clientsession_kwargs)
-        # pyright has trouble to understand that self.session is not None, since we raised at worst in the init
-        self.session = cast(aiohttp.ClientSession, self.session)
+        if self._has_been_opened and self._has_been_closed:
```
Comment (xiangyan99) [inline sdk/core/azure-core/azure/core/pipeline/transport/_aiohttp.py:132]: transport.close()
transport.open()
transport.open()   <-- we raise error here? why?

Comment (lmazuel) [reply sdk/core/azure-core/azure/core/pipeline/transport/_aiohttp.py]: As of now we won't, since sessions is not None, double open works fine

**Comment by xiangyan99** [comment]
Now in the case of 
open -> close -> open again,
we change the error type from `AttributeError` to `ValueError`, should we keep raising `AttributeError` and just update the error message?

**Comment by lmazuel** [comment]
> Now in the case of open -> close -> open again, we change the error type from `AttributeError` to `ValueError`, should we keep raising `AttributeError` and just update the error message?

To me, an `AttributeError` from a client like that is a bug, so we're not breaking working code, but fixing an unexpected error. I don't see a scenario where a customer will write code to expect an `AttributeError` and recover from it (but of course everything is a breaking change :)). So I still vote to keep `ValueError`, but of course I'm open to various opinions.

**Review by xiangyan99 — APPROVED** [review]
LGTM.

Unless our architects have concerns on the error type change.

**Comment by lmazuel** [comment]
> We should make sure we didn't break child clients/shared transports (e.g. storage container -> blob client)

We're passing the Storage tests, is that enough @johanste ?

File: sdk/core/azure-core/azure/core/pipeline/transport/_aiohttp.py [hunk 2]
```diff
@@ -126,26 +128,33 @@ async def __aexit__(
...
-            self.session = aiohttp.ClientSession(**clientsession_kwargs)
-        # pyright has trouble to understand that self.session is not None, since we raised at worst in the init
-        self.session = cast(aiohttp.ClientSession, self.session)
+        if self._has_been_opened and not self.session:
+            raise ValueError(
+                "HTTP transport has already been closed. "
+                "You may check if you're calling a function outside of the `async with` of your client creation, "
+                "or if you called `await close()` on your client already."
+            )
+        if not self.session:
```
Comment (annatisch) [inline sdk/core/azure-core/azure/core/pipeline/transport/_aiohttp.py:137]: There is still a small possibility that by the time get down to `await self.session.request()` that the transport has been closed by another thread - so I wouldn't be opposed to also adding a catch for the `AttributeError` to that call. But these changes definitely make the window for that race condition occurring much narrower, so probably not crucial.

**Review by annatisch — APPROVED** [review]

**Pull request merged**